components/feature/accounts/src/main/java/mozilla/components/feature/accounts/FxaSupportFeature.kt

```diff
@@ -0,0 +1,109 @@
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.sync.AuthType
+import mozilla.components.concept.sync.ConstellationState
+import mozilla.components.concept.sync.DeviceConstellation
+import mozilla.components.concept.sync.DeviceConstellationObserver
+import mozilla.components.concept.sync.OAuthAccount
+import mozilla.components.concept.sync.AccountObserver as SyncAccountObserver
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.support.base.log.logger.Logger
+
+internal const val PREFERENCE_NAME = "mozac_feature_accounts"
+internal const val LAST_VERIFIED = "last_verified_push_subscription"
+internal const val PERIODIC_INTERVAL_MILLISECONDS = 24 * 60 * 60 * 1000L // 24 hours
+
+class FxaSupportFeature(
+    context: Context,
+    accountManager: FxaAccountManager,
+    push: PushProcessor,
+    lifecycleOwner: LifecycleOwner = ProcessLifecycleOwner.get(),
+    autoPause: Boolean = false
+) {
+    init {
+        val constellationObserver = ConstellationObserver(context, push)
+
+        val accountObserver = AccountObserver(context, accountManager, constellationObserver, lifecycleOwner, autoPause)
+
+        accountManager.register(accountObserver)
+    }
+}
+
+internal class AccountObserver(
+    private val context: Context,
+    private val accountManager: FxaAccountManager,
+    private val observer: DeviceConstellationObserver,
+    private val lifecycleOwner: LifecycleOwner,
+    private val autoPause: Boolean
+) : SyncAccountObserver {
+    private val logger = Logger("AccountObserver")
+
+    override fun onAuthenticated(account: OAuthAccount, authType: AuthType) {
+        // Register observer when we have an existing account.
+        if (authType != AuthType.Existing) {
+            return
+        }
+
+        accountManager.withConstellation { constellation ->
+            constellation.registerDeviceObserver(observer, lifecycleOwner, autoPause)
+        }
+    }
+
+    override fun onLoggedOut() {
+        // Delete renewal pref.
+        preference(context).edit().remove(LAST_VERIFIED).apply()
+    }
+}
+
+internal class ConstellationObserver(
+    private val context: Context,
+    private val push: PushProcessor
+) : DeviceConstellationObserver {
+
+    private val logger = Logger("ConstellationObserver")
+    private var lastVerified: Long
+        get() = preference(context).getLong(LAST_VERIFIED, System.currentTimeMillis())
+        set(value) = preference(context).edit().putLong(LAST_VERIFIED, value).apply()
+
+    override fun onDevicesUpdate(constellation: ConstellationState) {
+        val updateSubscription = constellation.currentDevice?.subscriptionExpired ?: false
+
+        // If our subscription has not expired, we do nothing.
+        if (updateSubscription) {
+            return
+        }
+
+        // If our last check was recent (see: PERIODIC_INTERVAL_MILLISECONDS), we do nothing.
+        if (shouldVerifyNow().not()) {
```

Comment:
if we're told by our own device record that we have a bad subscription, we should `renewRegistration` regardless of the timestamps.